doc/modules/preprocessing.rst

```diff
@@ -455,47 +455,87 @@ Such features can be efficiently coded as integers, for instance
-numbers the continent and the last four the web browser.
-
-Note that, if there is a possibility that the training data might have missing categorical
-features, one has to explicitly set ``n_values``. For example,
-
-    >>> enc = preprocessing.OneHotEncoder(n_values=[2, 3, 4])
-    >>> # Note that there are missing categorical values for the 2nd and 3rd
-    >>> # features
-    >>> enc.fit([[1, 2, 3], [0, 2, 0]])  # doctest: +ELLIPSIS
-    OneHotEncoder(categorical_features='all', dtype=<... 'numpy.float64'>,
-           handle_unknown='error', n_values=[2, 3, 4], sparse=True)
-    >>> enc.transform([[1, 0, 0]]).toarray()
-    array([[ 0.,  1.,  1.,  0.,  0.,  1.,  0.,  0.,  0.]])
+  >>> enc = preprocessing.CategoricalEncoder()
+  >>> X = [['male', 'from US', 'uses Safari'], ['female', 'from Europe', 'uses Firefox']]
+  >>> enc.fit(X)  # doctest: +ELLIPSIS
+  CategoricalEncoder(categories='auto', dtype=<... 'numpy.float64'>,
+            encoding='onehot', handle_unknown='error')
+  >>> enc.transform([['female', 'from US', 'uses Safari'],
+  ...                ['male', 'from Europe', 'uses Safari']]).toarray()
+  array([[ 1.,  0.,  0.,  1.,  0.,  1.],
+         [ 0.,  1.,  1.,  0.,  0.,  1.]])
+
+By default, the values each feature can take is inferred automatically
+from the dataset and can be found in the ``categories_`` attribute::
+
+    >>> enc.categories_
+    [array(['female', 'male'], dtype=object), array(['from Europe', 'from US'], dtype=object), array(['uses Firefox', 'uses Safari'], dtype=object)]
+
+It is possible to specify this explicitly using the parameter ``categories``.
+There are two genders, four possible continents and four web browsers in our
+dataset::
+
+    >>> genders = ['female', 'male']
+    >>> locations = ['from Africa', 'from Asia', 'from Europe', 'from US']
+    >>> browsers = ['uses Chrome', 'uses Firefox', 'uses IE', 'uses Safari']
+    >>> enc = preprocessing.CategoricalEncoder(categories=[genders, locations, browsers])
+    >>> # Note that for there are missing categorical values for the 2nd and 3rd
+    >>> # feature
+    >>> X = [['male', 'from US', 'uses Safari'], ['female', 'from Europe', 'uses Firefox']]
+    >>> enc.fit(X) # doctest: +ELLIPSIS
+    CategoricalEncoder(categories=[...],
+              dtype=<... 'numpy.float64'>, encoding='onehot',
+              handle_unknown='error')
+    >>> enc.transform([['female', 'from Asia', 'uses Chrome']]).toarray()
+    array([[ 1.,  0.,  0.,  1.,  0.,  0.,  1.,  0.,  0.,  0.]])
+
+If there is a possibility that the training data might have missing categorical
+features, it can often be better to specify ``handle_unknown='ignore'`` instead
+of setting the ``categories`` manually as above. When
+``handle_unknown='ignore'`` is specified and unknown categories are encountered
+during transform, no error will be raised but the resulting one-hot encoded
+columns for this feature will be all zeros
```

Comment:
Really? If a category doesn't exist in training, how can it produce a zero column?? I thought this was the description for when specifying categories.

Reply:
OK, I agree this can be confusing, but you can read it both ways: also in the case of `handle_unknown='ignore'` you end up with all zero's, just some zero's less as with the manually specified categories.

Note that this is also how it is explained in the class docstring's explanation of `handle_unknown`

Reply:
On second thought: IMO it is correct how I stated it, and I don't think it can be read both ways:

- if you manually specify categories, and you encounter a category that didn't exist in training but was specified in the categories: you get a 1 for that category (so eg [0, 0, 0, 1]
- if you didn't specify categories but used `handle_unknown='ignore'`, and you encounter a category that didn't exist in training: you get all zerors (so eg [0, 0, 0])

So for me the above explanation is correct. It is not that easy to clearly word the idea of "a zero for each dummy column for that specific feature in that row", so if you have a better wording than "the resulting one-hot encoded columns for this feature will be all zeros", always welcome.

Or if the above is not clear, can you try to clarify your confusion with the text?

Reply:
You're right. I misread. Thanks for the explanation.